### PR TITLE
Fix IgnorePortInHostMatching always set to true, now determining it based on proxy and listenerport

### PR DESCRIPTION
**Pull request by heyanchun**
Fix IgnorePortInHostMatching always set to true, now determining it based on proxy and listenerport

**Please provide a description of this PR:**

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 1]
```diff
@@ -213,7 +213,7 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundHTTPRouteConfig(
 		VirtualHosts:                   virtualHosts,
 		ValidateClusters:               proto.BoolFalse,
 		MaxDirectResponseBodySizeBytes: istio_route.DefaultMaxDirectResponseBodySizeBytes,
-		IgnorePortInHostMatching:       true,
+		IgnorePortInHostMatching:       IgnorePortInHostMatching(listenerPort, node),
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/httproute.go:216]: It was intended to ignore port during host match for sidecar

Comment (heyanchun) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:216]: Not always. When listenerPort is 0, it will generate virtual host's domains with port, and now if sidecar ignore port during host matching, it can not match

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:216]: That sounds like a bug in the generation, not here?

Can you give more info on the exact setup that this fixes (and maybe add a test)?

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:216]: I kind of understand this bug. There is a inconsistency when listenerport = 0.

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 2]
```diff
@@ -615,10 +615,14 @@ func SidecarIgnorePort(node *model.Proxy) bool {
 	return !node.IsProxylessGrpc()
 }
 
+func IgnorePortInHostMatching(listenerPort int, node *model.Proxy) bool {
+	return listenerPort != 0 && SidecarIgnorePort(node)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/httproute.go:619]: for sidecar we may always set it true and no need check listenerport

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 1 above)
Comment (heyanchun) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:216]: > That sounds like a bug in the generation, not here?
>
> Can you give more info on the exact setup that this fixes (and maybe add a test)?

For example, when listenerPort = 0, sidecar will get dynamic route configs like
```
"dynamic_route_configs": [
{
  "name": "0",`
  "virtual_hosts": [
    "name": "a-b-c.svc.cluster.local:10000",
    "domains": [
        "a-b-c.svc.cluster.local:10000",
        "a-b-c:10000",
    ]
  ]
}]
```
When listenerPort = 10000, sidecar will get dynamic route configs like
```
"dynamic_route_configs": [
{
  "name": "10000",`
  "virtual_hosts": [
    "name": "a-b-c.svc.cluster.local:10000",
    "domains": [
        "a-b-c.svc.cluster.local",
        "a-b-c",
    ]
  ]
}]
```
If we set IgnorePortInHostMatching true, sidecar will match domains with "a-b-c.svc.cluster.local", not "a-b-c.svc.cluster.local:10000". It can not match when listenerPort = 0.
Further more, when listenerPort = 0 we could not generate domain "a-b-c.svc.cluster.local" and "a-b-c" in buildVirtualHost step because service a-b-c may has more than one port.

**Comment by howardjohn** [comment]
how does port 0 happen in real world?

**Comment by heyanchun** [comment]
> how does port 0 happen in real world?

1. see https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/httproute.go#L384
```
func BuildSidecarOutboundVirtualHosts(){
  // When generating RDS for ports created via the SidecarScope, we treat ports as HTTP proxy style ports
  // if ports protocol is HTTP_PROXY.
  if egressListener.IstioListener != nil && egressListener.IstioListener.Port != nil &&
    protocol.Parse(egressListener.IstioListener.Port.Protocol) == protocol.HTTP_PROXY {
    listenerPort = 0
  }
  ...
}
```
2. It also happens when we permanently set sidecar's static resources like
```
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        address: 0.0.0.0
        port_value: 12345
    filter_chains:
    - filters:
      - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          use_remote_address: true
          stat_prefix: envoy
          codec_type: AUTO
          rds:
            route_config_name: "0"  # it means routename = 0 and listenerport = 0
            config_source:
              ads: {}
          http_filters:
          - name: envoy.filters.http.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
```

File: pilot/pkg/networking/core/v1alpha3/httproute.go [hunk 3]
```diff
@@ -609,6 +609,18 @@ func getVirtualHostsForSniffedServicePort(vhosts []*route.VirtualHost, routeName
...
 }
 
+func IgnorePortInHostMatching(routeName string, listenerPort int, node *model.Proxy) bool {
+	// When generating RDS for ports created via the SidecarScope, we treat ports as HTTP proxy style ports
+	// if ports protocol is HTTP_PROXY.
+	egressListener := node.SidecarScope.GetEgressListenerForRDS(listenerPort, routeName)
+	if egressListener != nil && egressListener.IstioListener != nil && egressListener.IstioListener.Port != nil &&
+		protocol.Parse(egressListener.IstioListener.Port.Protocol) == protocol.HTTP_PROXY {
+		listenerPort = 0
+	}
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/httproute.go:619]: This seems inappropriate here.. 

Don't we already do it in BuildSidecarOutboundVirtualHosts?

We shouldn't have two places we do this. its expensive and makes the logic hard to follow

**Review by howardjohn — COMMENTED** [review]
Ok with the idea, but we need some e2e test coverage here ideally...

**Comment by heyanchun** [comment]
> Ok with the idea, but we need some e2e test coverage here ideally...

ok, thanks. I will add some test here

File: pilot/pkg/networking/core/v1alpha3/httproute.go (hunk 3 above)
Comment (heyanchun) [reply pilot/pkg/networking/core/v1alpha3/httproute.go:619]: yes, it is indeed repetitive code. I will fix it.
